### ISSUE TITLE
torch._C._LinAlgError on whiten_data on clifford group norm

### ISSUE THREAD

**Issue by liluo2** [pdearena/pdearena#29]
Hi! I encountered a problem when I was playing with clifford algebra.

On clifford U-net (just replace all the layers on UNet function of `oned_unet.py` with corresponding clifford layers), this error was raised:

`File "/home/liluo/code/pde-refiner/pdearena/models/pderefiner.py", line 136, in train_step                   
pred = self.model(x_in, time=k * self.time_multiplier, z=cond)                                                        
File "/home/liluo/anaconda3/envs/pderefiner/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl                                                                                                             
return self._call_impl(*args, **kwargs)                                                                               
File "/home/liluo/anaconda3/envs/pderefiner/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl                                                                                                                     return forward_call(*args, **kwargs)                                                                                  
File "/home/liluo/code/pde-refiner/pdearena/modules/conditioned/twod_cliffordunet.py", line 736, in forward               
x = m(x, emb)                                                                                                         
File "/home/liluo/anaconda3/envs/pderefiner/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl                                                                                                             
return self._call_impl(*args, **kwargs)                                                                              
File "/home/liluo/anaconda3/envs/pderefiner/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl                                                                                                                     return forward_call(*args, **kwargs)                                                                                  
File "/home/liluo/code/pde-refiner/pdearena/modules/conditioned/twod_cliffordunet.py", line 341, in forward               
x = self.res(x, emb)                                                                                                  
File "/home/liluo/anaconda3/envs/pderefiner/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl                                                                                                             
return self._call_impl(*args, **kwargs)                                                                               
File "/home/liluo/anaconda3/envs/pderefiner/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl                                                                                                                     return forward_call(*args, **kwargs)                                                                                  
File "/home/liluo/code/pde-refiner/pdearena/modules/conditioned/twod_cliffordunet.py", line 79, in forward                
h = self.conv1(self.activation(self.norm1(x))) 
File "/home/liluo/anaconda3/envs/pderefiner/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/liluo/anaconda3/envs/pderefiner/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/liluo/anaconda3/envs/pderefiner/lib/python3.9/site-packages/cliffordlayers/nn/modules/groupnorm.py", line 123, in forward
    return clifford_group_norm(
  File "/home/liluo/anaconda3/envs/pderefiner/lib/python3.9/site-packages/cliffordlayers/nn/functional/groupnorm.py", line 215, in clifford_group_norm
    return _instance_norm(
  File "/home/liluo/anaconda3/envs/pderefiner/lib/python3.9/site-packages/cliffordlayers/nn/functional/groupnorm.py", line 195, in _instance_norm
    x_norm = clifford_batch_norm(
  File "/home/liluo/anaconda3/envs/pderefiner/lib/python3.9/site-packages/cliffordlayers/nn/functional/batchnorm.py", line 178, in clifford_batch_norm
    x_norm = whiten_data(
  File "/home/liluo/anaconda3/envs/pderefiner/lib/python3.9/site-packages/cliffordlayers/nn/functional/batchnorm.py", line 73, in whiten_data
    U = torch.linalg.cholesky(cov + eye).mH
**torch._C._LinAlgError("linalg.cholesky: (Batch element 699): The factorization could not be completed because the input is not positive-definite (the leading minor of order 3 is not positive-definite).")`**

Here I ran another experiment and printed the cov out:

`cov = torch.tensor([
    [[ 1.9155e+10, -2.5920e+10,  2.5743e+09, -1.6588e+10],
     [-2.5920e+10,  3.7468e+10, -1.3118e+09,  2.5689e+10],
     [ 2.5743e+09, -1.3118e+09,  2.3489e+09,  7.2574e+08],
     [-1.6588e+10,  2.5689e+10,  7.2574e+08,  1.8771e+10]],
    [[ 9.8773e+10, -1.3027e+11,  1.6077e+10, -8.7328e+10],
     [-1.3027e+11,  1.8396e+11, -1.0372e+10,  1.3149e+11],
     [ 1.6077e+10, -1.0372e+10,  1.2281e+10,  3.3376e+08],
     [-8.7328e+10,  1.3149e+11,  3.3376e+08,  9.9116e+10]],
    [[ 4.2664e+09, -2.1247e+09, -4.0075e+09,  9.1318e+08],
     [-2.1247e+09,  1.3469e+09,  1.9276e+09, -7.5359e+08],
     [-4.0075e+09,  1.9276e+09,  3.7827e+09, -7.8728e+08],
     [ 9.1318e+08, -7.5359e+08, -7.8728e+08,  5.1068e+08]]])`

I found the second batch is **negative-definite** as its eigenvalue: `tensor([-9.9568e+05,  2.2260e+06,  2.3272e+10,  3.7086e+11])`. But I don't know why because cov is calculated via this line: `cov = torch.matmul(X, X.transpose(-1, -2)) / X.shape[-1]` (I verified this because X was also printed). This should imply that **cov is at least semi-positive definite.** I don't know if you guys can give me a little direction on how to solve this? 

Sorry to interupt you guys~ Looking forward to your reply!

Best regrad, 
Li Luo

**Comment by brandstetter-johannes** [comment]
Hi  @liluo2 - seems the error appears here: https://github.com/microsoft/cliffordlayers/blob/main/cliffordlayers/nn/functional/batchnorm.py#L70

This is unfortunate and we never encountered that. But you should use a CliffordFluidUNet which is the best model and uses a different normalization: https://github.com/microsoft/cliffordlayers/blob/main/tests/test_GCAFluidNet2d.py#L5

These are the nets from https://arxiv.org/abs/2302.06594.

**Comment by liluo2** [comment]
Thanks for your reply Johannes @brandstetter-johannes! I'll try it.

BTW, in a month ago I save the whole X matrix and tried to calculate it by indexing the problematic matrix. I found that, for example, if I indexed X by the problematic index first and then calculate , cholesky can be calculated. But if I calculated it within a batch, the error occured. I printed out the error:
`
[[0., 256.，-1536., 2048.],
256.，1536.，-2048.，-5120.],
[-1536.，-2048.，16384.，10240.]
[ 2048.，-5120.，10240.， 10240.]]，device='cuda:5', grad_fn=<SubBackward0>)
`
At the deep level of CliffordUNet. So I have a question here: **why would the feature matrix X become so large? If this makes any sense then why?** I also tried to resume checkpoint and found that as training went on the scale of this matrix became larger and larger. I think it is the large scale that causes the error/instabilty of the gpu calculation.

So I modified the line 78 of whiten_data [https://github.com/microsoft/cliffordlayers/blob/main/cliffordlayers/nn/functional/batchnorm.py#L68](url). eps is generated by 
identifying the max value of each batch, multiplies it by '1e-5'. Then the problem seems solved and the result seems nice. 

As for the model you provided, I haven't considered introducing any group equivariant property (for example those GCA model). I'll try to analyze whether these can be used. Also, different normalization may be favorable. 

Looking forward to your reply and Best wishes!

**Comment by brandstetter-johannes** [comment]
Hi @liluo2 would you mind opening a PR in the Clifford repo? We could iterate there, your fix seems important!

**Comment by liluo2** [comment]
Yeah I'll do it @brandstetter-johannes .
As training goes on, loss becomes NaN. I think it is due to the overflow of this level. Could you tell me some suggestions if you have some thoughts maybe like modifying or altering the model?

**Comment by liluo2** [comment]
After altering model to [https://github.com/microsoft/cliffordlayers/blob/main/tests/test_GCAFluidNet2d.py#L5](url) and maybe some hyperparameter tuning, the problem solved! Closing this issue now. If there are more issues just reopen it. Thanks Johannes!
